### ISSUE TITLE
Obsolete warnings

### ISSUE THREAD

**Issue by Ergus** [millejoh/emacs-ipython-notebook#800]
Using emacs 28 in the master branch I get this warnings that maybe should be fixed in order to keep the order up to date. As the new native compiler is a bit more strict with the checks.

```
Warning (comp): ein-utils.el:164:13: Warning: reference to free variable ‘it’ Disable showing Disable logging
Warning (comp): ein-utils.el:560:4: Warning: the function ‘ein:log’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-utils.el:293:6: Warning: the function ‘comint-carriage-motion’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-utils.el:163:10: Warning: the function ‘aif’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-core.el:95:6: Warning: the function ‘ein:log’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-core.el:85:37: Warning: the function ‘ein:$kernelspec-language’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-log.el:101:10: Warning: the function ‘ein:$websocket-ws’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-log.el:100:9: Warning: the function ‘websocket-get-debug-buffer-create’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-log.el:98:26: Warning: the function ‘ein:$kernel-websocket’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-log.el:97:23: Warning: the function ‘ein:get-kernel--notebook’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-websocket.el:91:27: Warning: the function ‘ein:log’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:52:13: Warning: assignment to free variable ‘comint-use-prompt-regexp’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:53:13: Warning: assignment to free variable ‘comint-prompt-regexp’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:54:13: Warning: assignment to free variable ‘comint-input-sender’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:55:13: Warning: assignment to free variable ‘comint-prompt-read-only’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:75:27: Warning: reference to free variable ‘it’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:87:28: Warning: reference to free variable ‘it’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:90:45: Warning: ‘(session (ein:ipdb-get-session kernel))’ is a malformed function Disable showing Disable logging
Warning (comp): ein-ipdb.el:90:45: Warning: Unused lexical argument `kernel' Disable showing Disable logging
Warning (comp): ein-ipdb.el:104:36: Warning: reference to free variable ‘session’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:97:25: Warning: reference to free variable ‘buffer’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:98:32: Warning: reference to free variable ‘proc’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:104:14: Warning: reference to free variable ‘prompt’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:109:7: Warning: reference to free variable ‘comint-input-sender-no-newline’ Disable showing Disable logging
Warning (comp): ein-ipdb.el:119:36: Warning: the function ‘comint-mode’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:98:19: Warning: the function ‘proc-live-p’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:105:33: Warning: the function ‘proc’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:105:38: Warning: the function ‘prompt’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:95:19: Warning: the function ‘buffer’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:94:6: Warning: the function ‘-when-let*’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:93:8: Warning: the function ‘ein:json-read-from-string’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:86:13: Warning: the function ‘ein:notebook-buffer’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:74:4: Warning: the function ‘awhen’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:71:6: Warning: the function ‘ein:websocket-send-stdin-channel’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:65:10: Warning: the function ‘comint-output-filter’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-ipdb.el:36:13: Warning: the function ‘ein:$kernel-kernel-id’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-query.el:102:52: Warning: reference to free variable ‘ein:force-sync’ Disable showing Disable logging
Warning (comp): ein-query.el:134:46: Warning: the function ‘make-ein:$kernelspec’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-output-area.el:34:1: Warning: defvar `ein:output-area-case-types' docstring wider than 80 characters Disable showing Disable logging
Warning (comp): ein-file.el:26:15: Warning: reference to free variable ‘ein:content-file-buffer--content’ Disable showing Disable logging
Warning (comp): ein-file.el:43:11: Warning: assignment to free variable ‘ein:content-file-buffer--content’ Disable showing Disable logging
Warning (comp): ein-file.el:58:35: Warning: reference to free variable ‘ein:content-file-buffer--content’ Disable showing Disable logging
Warning (comp): ein-file.el:59:4: Warning: the function ‘ein:content-save’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:58:35: Warning: the function ‘(setf ein:$content-raw-content)’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:52:41: Warning: the function ‘ein:$content-name’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:52:6: Warning: the function ‘ein:log’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:44:25: Warning: the function ‘ein:$content-raw-content’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:42:66: Warning: the function ‘ein:$content-path’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:41:66: Warning: the function ‘ein:$content-url-or-port’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:38:4: Warning: the function ‘ein:content-query-contents’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:37:48: Warning: the function ‘ein:notebooklist-ask-path’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:37:17: Warning: the function ‘ein:notebooklist-parse-nbpath’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-file.el:26:2: Warning: the function ‘ein:deflocal’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-contents-api.el:47:1: Warning: custom-declare-variable `ein:content-query-max-branch' docstring wider than 80 characters Disable showing Disable logging
Warning (comp): ein-notebooklist.el:528:18: Warning: the function ‘ein:notebook-get-opened-notebook’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-notebooklist.el:396:9: Warning: the function ‘ein:kernel-delete-session’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-notebooklist.el:381:8: Warning: the function ‘ein:notebook-opened-notebooks’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-notebooklist.el:367:23: Warning: the function ‘ein:notebook-rename-command’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-notebooklist.el:366:42: Warning: the function ‘ein:notebook-buffer’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-notebooklist.el:332:6: Warning: the function ‘ein:notebook-open’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-notebooklist.el:305:24: Warning: the function ‘ein:list-available-kernels’ is not known to be defined. Disable showing Disable logging
Warning (comp): poly-ein.el:45:46: Warning: reference to free variable ‘poly-ein-mode’ Disable showing Disable logging
Warning (comp): poly-ein.el:65:9: Warning: reference to free variable ‘poly-ein-mode’ Disable showing Disable logging
Warning (comp): poly-ein.el:95:13: Warning: reference to free variable ‘poly-ein-mode’ Disable showing Disable logging
Warning (comp): poly-ein.el:313:14: Warning: reference to free variable ‘poly-ein-mode’ Disable showing Disable logging
Warning (comp): poly-ein.el:414:14: Warning: reference to free variable ‘desktop-minor-mode-table’ Disable showing Disable logging
Warning (comp): poly-ein.el:414:14: Warning: assignment to free variable ‘desktop-minor-mode-table’ Disable showing Disable logging
Warning (comp): poly-ein.el:380:18: Warning: the function ‘ido-to-end’ is not known to be defined. Disable showing Disable logging
Warning (comp): poly-ein.el:238:22: Warning: the function ‘ein:worksheet-get-current-cell’ is not known to be defined. Disable showing Disable logging
Warning (comp): poly-ein.el:198:50: Warning: the function ‘ein:notebook-buffer-list’ is not known to be defined. Disable showing Disable logging
Warning (comp): poly-ein.el:125:6: Warning: the function ‘poly-ein--syntax-propertize’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-worksheet.el:45:1: Warning: defconst `ein:worksheet--max-buffer-undo-list-length' docstring wider than 80 characters Disable showing Disable logging
Warning (comp): ein-worksheet.el:215:42: Warning: reference to free variable ‘ein:%worksheet%’ Disable showing Disable logging
Warning (comp): ein-traceback.el:133:24: Warning: the function ‘ein:worksheet-get-cells’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:46:33: Warning: reference to free variable ‘it’ Disable showing Disable logging
Warning (comp): ein-python-send.el:43:17: Warning: assignment to free variable ‘ein:%notebook%’ Disable showing Disable logging
Warning (comp): ein-python-send.el:133:25: Warning: reference to free variable ‘python-indent-offset’ Disable showing Disable logging
Warning (comp): ein-python-send.el:156:17: Warning: reference to free variable ‘python-mode-map’ Disable showing Disable logging
Warning (comp): ein-python-send.el:143:8: Warning: the function ‘python-nav-forward-statement’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:118:28: Warning: the function ‘python-info-current-line-comment-p’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:112:14: Warning: the function ‘python-nav-end-of-block’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:93:4: Warning: the function ‘python-nav-beginning-of-statement’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:82:37: Warning: the function ‘ein:get-kernel’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:82:6: Warning: the function ‘ein:shared-output-eval-string’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:54:22: Warning: the function ‘python-shell-buffer-substring’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:42:39: Warning: the function ‘ein:$notebook-notebook-name’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:40:17: Warning: the function ‘ein:completing-read’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:38:11: Warning: the function ‘ein:notebook-opened-notebooks’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-python-send.el:38:6: Warning: the function ‘aif’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-notebook.el:316:1: Warning: defalias `ein:notebook-reconnect-kernel' docstring wider than 80 characters Disable showing Disable logging
Warning (comp): ein-notebook.el:995:55: Warning: the function ‘ein:jupyter-server-stop’ is not known to be defined. Disable showing Disable logging
Warning (comp): ein-jupyter.el:45:1: Warning: custom-declare-variable `ein:jupyter-docker-mount-point' docstring wider than 80 characters Disable showing Disable logging
Warning (comp): ein-jupyter.el:62:1: Warning: custom-declare-variable `ein:jupyter-server-command' docstring wider than 80 characters Disable showing Disable logging
Warning (comp): ein-jupyter.el:153:11: Warning: defsubst ‘ein:jupyter-server-process’ was used before it was defined Disable showing Disable logging
Warning (comp): ein-process.el:154:11: Warning: defsubst ‘ein:process-url-or-port’ was used before it was defined Disable showing Disable logging

```

**Comment by Ergus** [comment]
> From the README:
> […](#)
> Note EIN is tested only for released GNU Emacs versions 25.1 and later. Pre-release versions will not work.

Hi @dickmao  AFAIU 28 (next release) is LATER than 25.1 it is actually the latest next release.

**Comment by dickmao** [comment]
JAM>  Note EIN is tested only for released GNU Emacs versions 25.1 and later. Pre-release versions will
JAM>  not work.

Your observation that 28 is numerically greater than 25 does not make the
above statement any less true.

"Premature optimization is the root of all evil."
